### PR TITLE
Fixed crash on circular local type arguments when outer ones are present too

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/59062

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot test it

**Review by andrewbranch — APPROVED** [review]

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -16298,7 +16298,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     function getTypeArguments(type: TypeReference): readonly Type[] {
         if (!type.resolvedTypeArguments) {
             if (!pushTypeResolution(type, TypeSystemPropertyName.ResolvedTypeArguments)) {
-                return type.target.localTypeParameters?.map(() => errorType) || emptyArray;
+                return concatenate(type.target.outerTypeParameters, type.target.localTypeParameters?.map(() => errorType) || emptyArray);
```
Comment (jakebailey) [inline src/compiler/checker.ts:16301]: No need for the `emptyArray` bit given the parameter accepts undefined.

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
Approved, though I'd also like the nits from Jake fixed up.

File: src/compiler/checker.ts (hunk 1 above)
Comment (Andarist) [reply src/compiler/checker.ts:16301]: it's actually kinda needed so `concatenate` doesn't return `undefined`. If I remove it get type errors here

Comment (jakebailey) [reply src/compiler/checker.ts:16301]: Could stick it on the outside, maybe...

Comment (Andarist) [reply src/compiler/checker.ts:16301]: I'm not sure what do you mean

Comment (jakebailey) [reply src/compiler/checker.ts:16301]: As in `return concatenate(a, b) ?? emptyArray`

But maybe that's worse. Can't recall if this helper will make a spurious copy on an empty input array or not

**Comment by Andarist** [comment]
thanks @gabritto !

**Pull request merged**